torch/autograd/gradcheck.py

```diff
@@ -258,14 +257,21 @@ def fail_test(msg):
 
     # Make sure that gradients are saved for at least one input
     any_input_requiring_grad = False
-    for inp in tupled_inputs:
+    for idx, inp in enumerate(tupled_inputs):
         if isinstance(inp, torch.Tensor) and inp.requires_grad:
             if not (inp.dtype == torch.float64 or inp.dtype == torch.complex128):
                 warnings.warn(
-                    'At least one of the inputs that requires gradient '
-                    'is not of double precision floating point or complex. '
+                    'The {}th input requires gradient and '
+                    'is not a double precision floating point or complex. '
                     'This check will likely fail if all the inputs are '
                     'not of double precision floating point or complex. ')
+            content = inp._values() if inp.is_sparse else inp
+            if content.layout is not torch._mkldnn and any([s == 0 for s in content.stride()]):
```

Comment:
No... you should also check if size > 1. If size <= 1 and stride = 0 it is okay. Also didn’t we have some preliminary code for checking overlapping indices?

Reply:
here https://github.com/pytorch/pytorch/blob/959afe0726527c632fcdd09c3ef13fd45e772421/aten/src/ATen/templates/TensorBody.h#L213

Reply:
hmm also https://github.com/pytorch/pytorch/blob/2af64ba3ede8906c4b48e1d0988665520ad03c9d/aten/src/ATen/MemoryOverlap.cpp and https://github.com/pytorch/pytorch/blob/3def76583af948bf7d7d9eb8a794d3406f2bbed2/aten/src/ATen/cuda/detail/IndexUtils.cu#L37

we should really merge them... (cf https://github.com/pytorch/pytorch/pull/23586)

Reply:
Yes we have a few of these but none of them was exposed to python... So I went with the simple check here.

I guess Tensors with `nelement == 1` can be excluded indeed. Not sure if we have many of them with stride 0 though...

Reply:
The check in IndexUtils incorrectly computes the extent of the dimension https://github.com/pytorch/pytorch/blob/3def76583af948bf7d7d9eb8a794d3406f2bbed2/aten/src/ATen/cuda/detail/IndexUtils.cu#L64-L68
but we did not fix it because looks like nothing is using it. The extent of `i`-th sorted dimension is `(size-1)*stride+extent[i-1]`, not `(size-1)*stride + 1` as implied here. 
But for autograd this maybe too involved, and just checking for 0 strides will weed out 99% erroneous cases.

Reply:
Okay, but I don't think the check should be overly conservative (erroring out when it could be fine). It should detect a subset of error cases and then error out. So how about let's add the `size` check and a TODO note to use existing functionalities when they are ready?

Reply:
That sounds good. Opened a PR to do that: https://github.com/pytorch/pytorch/pull/38877